tests/result.cpp

```diff
@@ -0,0 +1,94 @@
+#include "gtest/gtest.h"
+
+#include "util/result.h"
+
+namespace bpftrace::test::result {
+
+class NotOK : public ErrorInfo<NotOK> {
+public:
+  static char ID;
+  void log(llvm::raw_ostream &OS) const override
+  {
+    OS << "Not OK.";
+  }
+};
+
+class ReallyNotOK : public ErrorInfo<ReallyNotOK> {
+public:
+  static char ID;
+  void log(llvm::raw_ostream &OS) const override
+  {
+    OS << "Really not OK.";
+  }
+};
+
+char NotOK::ID;
+char ReallyNotOK::ID;
+
+static Result<> alwaysOK()
+{
+  return OK();
+}
+
+static Result<> alwaysNotOK()
+{
+  return make_error<NotOK>();
+}
+
+static Result<bool> maybeOK(bool ok, bool really_bad)
+{
+  if (!ok) {
+    if (really_bad)
+      return make_error<ReallyNotOK>();
+    return make_error<NotOK>();
+  }
+  return really_bad; // Arbitrary value.
+}
+
+TEST(result, okay)
+{
+  auto ok = alwaysOK();
+  EXPECT_TRUE(bool(ok));
+}
+
+TEST(result, not_okay)
+{
+  auto ok = alwaysNotOK();
+  EXPECT_FALSE(bool(ok));
+}
+
+TEST(result, values)
+{
+  auto ok = maybeOK(true, false);
+  EXPECT_TRUE(bool(ok));
+  EXPECT_FALSE(*ok);
+  ok = maybeOK(true, true);
+  EXPECT_TRUE(bool(ok));
+  EXPECT_TRUE(*ok);
+}
+
+TEST(result, handle_err_found)
+{
+  auto ok = maybeOK(false, false);
+  if (!ok) {
+    auto nowOk = handleErrors(std::move(ok), [](const NotOK &) {});
+    EXPECT_TRUE(bool(nowOk)); // Should now be fine.
+  }
+}
+
+TEST(result, handle_err_missing)
+{
+  auto ok = maybeOK(false, true);
+  if (!ok) {
+    auto nowOk = handleErrors(std::move(ok), [](const NotOK &) {});
```

Comment:
So the result of this is false because the lambda is accepting a `NotOK` instead of a `ReallyNotOK` right? Just sanity checking myself.

Reply:
Correct, we are handling `NotOK` but not `ReallyNotOK`.

Reply:
You could provide a second handler (or third, or fourth, etc.) if you wanted to. But I elected *not* provide an API for `handleAllErrors` since I think it's fair to require that no caller is ever *sure* that they handle every error case.

(You can also check it and just print it and ignore it or whatever. But `handleErrors` lets you unpack the actual error information.)